builtin/logical/totp/path_roles.go

```diff
@@ -0,0 +1,342 @@
+				Default:     0,
+				Description: `The number of delay periods that are allowed when validating a TOTP token.`,
+			},
+
+			"qr_size": {
+				Type:        framework.TypeInt,
+				Default:     200,
+				Description: `The pixel size of the generated square QR code.`,
+			},
+		},
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.ReadOperation:   b.pathRoleRead,
+			logical.UpdateOperation: b.pathRoleCreate,
+			logical.DeleteOperation: b.pathRoleDelete,
+		},
+
+		HelpSynopsis:    pathRoleHelpSyn,
+		HelpDescription: pathRoleHelpDesc,
+	}
+}
+
+func (b *backend) Role(s logical.Storage, n string) (*roleEntry, error) {
+	entry, err := s.Get("key/" + n)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	var result roleEntry
+	if err := entry.DecodeJSON(&result); err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+func (b *backend) pathRoleDelete(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	err := req.Storage.Delete("key/" + data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (b *backend) pathRoleRead(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	role, err := b.Role(req.Storage, data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return nil, nil
+	}
+
+	switch role.Generate {
+	case true:
+		key_object, err := otplib.NewKeyFromURL(role.URL)
+
+		if err != nil {
+			return logical.ErrorResponse("An error occured while generating a Key object."), nil
+		}
+
+		barcode, err := key_object.Image(role.QRSize, role.QRSize)
+
+		if err != nil {
+			return logical.ErrorResponse("An error occured while generating a QR code image."), nil
+		}
+
+		var buff bytes.Buffer
+		png.Encode(&buff, barcode)
+		b64Barcode := base64.StdEncoding.EncodeToString(buff.Bytes())
+		return &logical.Response{
+			Data: map[string]interface{}{
+				"url":     key_object.String(),
+				"barcode": b64Barcode,
+				"skew":    role.Skew,
+				"qr_size": role.QRSize,
+			},
+		}, nil
+	default:
+		// Translate algorithm back to string
+		algorithm := role.Algorithm.String()
+
+		// Return values of key
+		return &logical.Response{
+			Data: map[string]interface{}{
+				"issuer":       role.Issuer,
+				"account_name": role.AccountName,
+				"period":       role.Period,
+				"algorithm":    algorithm,
+				"digits":       role.Digits,
+			},
+		}, nil
+	}
+}
+
+func (b *backend) pathRoleList(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	entries, err := req.Storage.List("key/")
+	if err != nil {
+		return nil, err
+	}
+
+	return logical.ListResponse(entries), nil
+}
+
+func (b *backend) pathRoleCreate(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	name := data.Get("name").(string)
+	generate := data.Get("generate").(bool)
+	key_string := data.Get("key").(string)
```

Comment:
Please follow Vault style and use camelcased var names like keyString, accountName, qrSize, keySize.